### PR TITLE
Api - Allocation request type refactor

### PR DESCRIPTION
**Pull request by HansDahle**
- [ ] New feature
- [ ] Bug fix
- [x] High impact

**Description of work:**
Refactored request type to simply allocation for normal/direct/jv and moved to sub type. The subtype is resolved automatically if not provided. 

**Testing:**
- [ ] Can be tested
- [ ] Automatic tests created / updated
- [ ] Local tests are passing

TBD

**Checklist:**
- [ ] Considered automated tests
- [ ] Considered updating specification / documentation
- [ ] Considered work items 
- [ ] Considered security
- [ ] Performed developer testing
- [ ] Checklist finalized / ready for review

### CODE REVIEW

File: src/backend/api/Fusion.Resources.Database/Migrations/20210317154326_Added subtype to request.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+namespace Fusion.Resources.Database.Migrations
+{
+    public partial class Addedsubtypetorequest : Migration
+    {
+        protected override void Up(MigrationBuilder migrationBuilder)
+        {
+            migrationBuilder.AddColumn<string>(
+                name: "SubType",
+                table: "ResourceAllocationRequests",
+                type: "nvarchar(max)",
```
Comment (loldot) [inline src/backend/api/Fusion.Resources.Database/Migrations/20210317154326_Added subtype to request.cs:12]: I guess this is nitpicking, and a problem many other places, but we should probably avoid using nvarchar(max) for short, known(ish) length strings as nvarchar(max) values are stored outside the row in the database and requires more reads for each row. This is especially the case if we want to index the field.

File: src/backend/api/Fusion.Resources.Logic/Requests/Commands/ResourceAllocationRequest/Allocation/CanApproveStepHandler.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+                //var request = await dbContext.ResourceAllocationRequests.FirstAsync(r => r.Id == notification.RequestId);
+
+                switch (notification.CurrentStepId)
+                {
+                    case "proposed":
+
+                        break;
+                }
+
+                return Task.CompletedTask;
```
Comment (loldot) [inline src/backend/api/Fusion.Resources.Logic/Requests/Commands/ResourceAllocationRequest/Allocation/CanApproveStepHandler.cs:35]: Should this be implemented?

File: src/backend/api/Fusion.Resources.Logic/Requests/Commands/ResourceAllocationRequest/Allocation/InitializedHandler.cs [hunk 3]
```diff
@@ -28,24 +26,45 @@ public InitializedHandler(ResourcesDbContext dbContext, IMediator mediator)
...
 
-                    var workflow = new InternalRequestDirectWorkflowV1(initiatedBy);
                     dbContext.Workflows.Add(workflow.CreateDatabaseEntity(notification.RequestId, DbRequestType.InternalRequest));
 
                     request.LastActivity = DateTime.UtcNow;
-                    request.State.State = InternalRequestDirectWorkflowV1.CREATED;
+                    request.State.State = AllocationDirectWorkflowV1.CREATED;
+
+                    // Try to resolve the default assigned department
+                    request.AssignedDepartment = await mediator.Send(new Queries.ResolveResponsibleDepartment(request.Id));
```
Comment (loldot) [inline src/backend/api/Fusion.Resources.Logic/Requests/Commands/ResourceAllocationRequest/Allocation/InitializedHandler.cs:59]: Would this not overwrite the assigned department if it is sent through the API, or are we deprecating that option?

File: src/backend/api/Fusion.Resources.Database/Migrations/20210317154326_Added subtype to request.cs (hunk 1 above)
Comment (HansDahle) [reply src/backend/api/Fusion.Resources.Database/Migrations/20210317154326_Added subtype to request.cs:12]: Yeh, it's not even nitpicking.. forgot about this.

File: src/backend/api/Fusion.Resources.Logic/Requests/Commands/ResourceAllocationRequest/Allocation/InitializedHandler.cs (hunk 3 above)
Comment (HansDahle) [reply src/backend/api/Fusion.Resources.Logic/Requests/Commands/ResourceAllocationRequest/Allocation/InitializedHandler.cs:59]: It is sort of deprecated, but could add a nullcheck here, so we do not blindly overwrite..

File: src/backend/api/Fusion.Resources.Database/Migrations/20210317154326_Added subtype to request.cs (hunk 1 above)
Comment (HansDahle) [reply src/backend/api/Fusion.Resources.Database/Migrations/20210317154326_Added subtype to request.cs:12]: I think we could probably do that in a separate pull request, for many different entities

File: src/backend/api/Fusion.Resources.Logic/Requests/Commands/ResourceAllocationRequest/Allocation/CanApproveStepHandler.cs (hunk 2 above)
Comment (HansDahle) [reply src/backend/api/Fusion.Resources.Logic/Requests/Commands/ResourceAllocationRequest/Allocation/CanApproveStepHandler.cs]: Added as placeholder, will implement when rules are more known.

**Review by loldot — APPROVED** [review]
💯

**Pull request merged**